### PR TITLE
Explain: Add dataflow information in optimized plan

### PR DESCRIPTION
**Pull request by wangandi**
Something I decided to spend my Friday doing because it's getting too annoying not to be able to see
* what a plan looks like with unmaterialized views inlined.
* Whether linear operators were properly pushed down to a source.

~~Based on [this comment](https://github.com/MaterializeInc/materialize/pull/5029#issuecomment-747768869), I have made the syntax to show the whole dataflow for a query/view look like this:
`EXPLAIN DATAFLOW (PLAN)? FOR ...`~~ `EXPLAIN OPTIMIZED PLAN...` now returns the plan after unmaterialized views have been inlined + information about linear operators that have been pushed down to sources.
Feel free to suggest alternate syntaxes.

The format for the dataflow explanation looks like this (see also test/testdrive/source-linear-operators.td in the diffs):
```
-- A source only shows up in the dataflow explanation if there is a linear operator to apply against the source
Source name (id)
| Filter ...
| Project

-- An intermediate view that will be built in the dataflow.
-- Realistically, no intermediate views show up in any dataflow explanation because inlining always kicks in.
View name (id):
%0 = 
...

-- The view/one-off select query being explained. If it is a view being explained, I prefer "Query: " 
-- to be replaced with the view name + id,but `Plan::Explain` does not include that information.
Query:
%0 = 
...
```
If there is only one view in the dataflow after inlining, and there are no sources with linear operators, ~~then the format of the dataflow explanation is the same as the format for `EXPLAIN PLAN ...`~~ then the format of the explanation is the same as it has always been. (see also test/sqllogictest/transform/dataflow.slt in the diffs).

Feel free to suggest alternate formats.

TODO:
*  Replace "Query: " with the view name + id if a view is being explained.
* Release notes and docs will be for another PR.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6173)

### CODE REVIEW

**Comment by frankmcsherry** [comment]
Drive by comment, but:
 > -- Realistically, no intermediate views show up in any dataflow explanation because inlining always kicks in.

At the moment we only inline if the output of a view is used only once. We could create `Let`/`Get` bindings, but we don't at the moment.

**Comment by benesch** [comment]
Exposing this information seems great! I don't quite follow the syntax additions though. When would you ever _not_ want to see this information? Can't we just include it by default if you type `CREATE OPTIMIZED PLAN`? I share your opinion that I saw somewhere about `EXPLAIN DATAFLOW` implying "explain the [hypothetical] LIR", not "explain `DataflowDesc`". 

Are there additional complexities here I'm not seeing?

**Comment by wangandi** [comment]
> Drive by comment, but:
> 
> > -- Realistically, no intermediate views show up in any dataflow explanation because inlining always kicks in.
> 
> At the moment we only inline if the output of a view is used only once. We could create `Let`/`Get` bindings, but we don't at the moment.

The comment in inline_views reads as follows:
```
// A view can / should be in-lined in another view if it is only used by
// one subsequent view. If there are two distinct views that have not
// themselves been merged, then too bad and it doesn't get inlined.
```
Currently, we only export one index or sink per dataflow, which corresponds to the last view in `objects_to_build`. If the second-to-last view is not used by the last view, then it is being created pointlessly since there are no other exports to pipe it to. Thus, realistically, the second-to-last view is always used by the last view, and they get merged together. Via induction, realistically inlining always kicks in.

If you are able to construct a sequence of unmaterialized views such that they don't all get inlined, do share it with me so I can add it as a test.

**Comment by frankmcsherry** [comment]
If by "realistically" you meant "currently" then I agree.

**Comment by wangandi** [comment]
> If by "realistically" you meant "currently" then I agree.

In my mind, "currently" is too vague because there is what our code currently can support vs. what currently runs. So I distinguish between those two concepts by calling them "theoretically" and "realistically" respectively. But as long as we're on the same page it's fine, especially since the term "realistically" is not in the doc comments in the PR.

> Exposing this information seems great! I don't quite follow the syntax additions though. When would you ever _not_ want to see this information? Can't we just include it by default if you type `CREATE OPTIMIZED PLAN`? I share your opinion that I saw somewhere about `EXPLAIN DATAFLOW` implying "explain the [hypothetical] LIR", not "explain `DataflowDesc`".
> 
> Are there additional complexities here I'm not seeing?

The comment you're talking about is [this one](https://github.com/MaterializeInc/materialize/pull/5029#issuecomment-747777718).

I think just including this information by default if you type `EXPLAIN OPTIMIZED PLAN` is a great idea. Though, I went through the exercise of checking if there are additional complexities to convince myself that including the information by default is ok, and I am leaving my analysis below for the sake of decision documentation. Let's refer to "not including the information" as "OPTIMIZED PLAN on main" and "including the information" as "OPTIMIZED PLAN in PR". 

1) Does "OPTIMIZED PLAN on main" capture intermediate state that is useful to look at? -- Currently a dataflow is optimized like this:
     1) Optimize each individual view.
     2) Inline views.
     3) Optimize the combined view.
     4) Push predicates/projections down to sources.
However, we have a TODO to remove step (i). So for now "OPTIMIZED PLAN on main" does capture intermediate state, but it will not in the future, and I think it is better for our `EXPLAIN` syntax to reflect the future. Also, what "OPTIMIZED PLAN on main" captures is not that useful since it is not a boundary within our stack like what "DECORRELATED PLAN" captures.
2) The `OPTIMIZED PLAN on main`  is the same as `OPTIMIZED PLAN on PR` if all the sources/views being depended on were materialized. Would it be bad if users lost the ability to see what a plan looks like if all dependent sources/views were materialized without actual materializing those sources and views? -- Currently, we make the user create actual indexes to see what the optimized plan looks like with those indexes, and that is ok. So it should be ok to have the user materialize dependent views to see what the plan looks like with dependent views materialized.
3) Would switching to "OPTIMIZED PLAN in PR" make it harder to identify which view a bad line in a plan comes from? -- If materialized view v2 depends on unmaterialized view v1, we should be able to determine whether a line in the explanation of `v2` resulted from the definition`v1` or `v2` by checking if the line is also in the explanation of `v1`.

File: test/testdrive/source-linear-operators.td [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+> CREATE SOURCE data
+  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-data-${testdrive.seed}'
+  FORMAT AVRO USING SCHEMA '${schema}'
+
+> CREATE VIEW data_view as SELECT * from data
+
+$ set-regex match=u\d+ replacement=UID
+
+> EXPLAIN PLAN FOR SELECT * from data_view where a = 1
+"Source materialize.public.data (UID):\n| Filter (#0 = 1)\n| Project (#0, #1)\n\nQuery:\n%0 =\n| Get materialize.public.data (UID)\n| Filter (#0 = 1)\n"
```
Comment (benesch) [inline test/testdrive/source-linear-operators.td:38]: Oof, that is thoroughly impossible to read. Sad, but seems fine for now. I wonder if @philip-stoev has thoughts on making this better.

**Review by benesch — COMMENTED** [review]
I'm loosely LGTM, in that I'm a little out of date on the dataflow stuff so I can't do the most thorough review. I don't love the `GlobalId::Explain` variant—these GlobalIds are proving to be a bad abstraction—but I don't have anything better to offer nor a full grasp of the complexities involved. Assuming Frank doesn't have philosophical objections I think you should merge this!

**Pull request merged**